### ISSUE TITLE
ValueError

### ISSUE THREAD

**Issue by asangphukieo** [mxmlnkn/pragzip#14]
Hi, 
I try to use pragzip to count total number of line in large VCF file and found this error in some outputs,

Traceback (most recent call last):
  File "/home/sangphukieoa/.local/bin/pragzip", line 8, in <module>
    sys.exit(cli())
  File "pragzip.pyx", line 254, in pragzip.cli
ValueError: Failed to decode deflate block at 23409459200 B 0 b because of: The backreferenced distance lies outside the window buffer!

My command is pragzip -d -c -P 0 vcf_file.dose.vcf.gz | wc -l
I use Slurm to run this command in parallel.

Many Thanks in advance.

**Comment by asangphukieo** [comment]
It is general text file with gzip. But in my case the number of column is more than 10000, it is larger than the general VCF file. I am not sure it is available online or not.

The size of the file is larger than 30 GB. 23409459200 B by chance or not, I am not sure, but I can check. I use CPU core in range 8 to 32. and I will try -v option , and will let you know the result.

Many thanks,

**Comment by asangphukieo** [comment]
It is general text file with gzip. But in my case the number of column is more than 10000, it is larger than the general VCF file. I am not sure it is available online or not.

The size of the file is larger than 30 GB. 23409459200 B by chance or not, I am not sure, but I can check. I use CPU core in range 8 to 32. and I will try -v option , and will let you know the result.

Many thanks,

**Comment by mxmlnkn** [comment]
I tried it with a 2.7 GB VCF file from [here](http://db.systemsbiology.net/kaviar/) and could not reproduce a problem. I have also recompressed it with igzip, pigz, gzip, and bgzip with all their compression levels for a total of 33 different versions encompassing 131 GiB and I could not reproduce your observed error. I am decompressing with 24 cores. (I did observe some other pecularities, though, such as `igzip -0` compressing everything into a single 2 GB long fixed Huffman compressed deflate block, which bloats the memory usage of pragzip in an unbounded manner given a large enough file and which makes parallelization impossible with the current approach)

I have further ideas for debugging:

 - Could you try and see whether the issue occurs with `pragzip -P 1`? This determines whether it is a problem with parallelization or with decompression in general.
 - Could you try `gzip --test vcf_file.dose.vcf.gz`? Maybe the file is broken.
 - Could you send me the output file of `dd if=vcf_file.dose.vcf.gz iflag=skip_bytes,count_bytes skip=$(( 23409459200 - 128 * 1024 ) count=$(( 256 * 1024 )) > extracted-chunk.dat`. This would give me 128 KiB of data before and after the faulty offset so that I can look into it locally. Of course, only if it is not private data.

**Comment by asangphukieo** [comment]
Thank you for your prompt response

- The number 23409459200 B  seem to be by chance and the error did not happen with single CPU run.

- This is the full error from -v option.

[10:17:31.243][140072751982400] [BlockFetcher::~BlockFetcher]
   Parallelization                   : 8
   Cache
       Hits                          : 0
       Misses                        : 9822
       Unused Entries                : 5579
       Maximum Fill Size             : 1
       Capacity                      : 16
   Prefetch Cache
       Hits                          : 4242
       Misses                        : 5580
       Unused Entries                : 0
       Prefetch Queue Hit            : 1337
       Maximum Fill Size             : 12
       Capacity                      : 16
   Cache Hit Rate                    : 49.9955 %
   Useless Prefetches                : 0 %
   Access Patterns
       Total Accesses                : 11159
       Duplicate Block Accesses      : 5580
       Sequential Block Accesses     : 5579
       Block Seeks Back              : 0
       Block Seeks Forward           : 0
   Blocks
       Total Existing                : >=5580
       Total Fetched                 : 5588
       Prefetched                    : 5587
       Fetched On-demand             : 1
   Prefetch Stall by BlockFinder     : 105
   Time spent in:
       bzip2::readBlockData          : 0 s
       decodeBlock                   : 11184.5 s
       std::future::get              : 112.93 s
       get                           : 548.798 s
   Thread Pool Utilization:
       Total Real Decode Duration    : 1753.34 s
       Theoretical Optimal Duration  : 1398.06 s
       Pool Efficiency (Fill Factor) : 79.7369 %
Traceback (most recent call last):
  File "/home/sangphukieoa/.local/bin/pragzip", line 8, in <module>
    sys.exit(cli())
  File "pragzip.pyx", line 254, in pragzip.cli
ValueError: Failed to read deflate block header at offset 23408410624 B 0 b (position after trying: 23408410625 B 0 b: Assumed padding seems to con
tain some kind of data!
 

-  Command gzip --test vcf_file.dose.vcf.gz works well.
- And sorry I cannot provide this dataset, it is private data.

**Comment by asangphukieo** [comment]
1. Please see see the log file of pragzip --analyze
[pragzip-analyze.log](https://github.com/mxmlnkn/pragzip/files/10939649/pragzip-analyze.log)

2. I will try the develop branch version soon, and let you know the result.
3. The vcf.gz file is generated by webserver https://imputation.biodatacatalyst.nhlbi.nih.gov/ , I have no idea which program they use to generate this file, bgzip is possible.
4. it is Linux x86_64, but I use Slurm to run the task with parallelization on HPC cluster.

Best,

**Comment by mxmlnkn** [comment]
>     3. The vcf.gz file is generated by webserver https://imputation.biodatacatalyst.nhlbi.nih.gov/ , I have no idea which program they use to generate this file, bgzip is possible.

I took a look at it and they even have their server code on Github and it also is documented that they use `bgzip`. I could also confirm this from the output of `--analyze`. I don't want to turn away one of my first users ;) but this means that you probably get a similar performance as `pragzip` with `bgzip`.

I'll try to compress some further files with bgzip to see whether I can reproduce a problem. Based on the output of `--analyze`, it looks like it crashes 461 MiB before the and of the compressed stream. Morever, because it is bgzip, the fast-path for BGZF files should be used, i.e., it is even more surprising that it doesn't work because I don't even have to look for deflate block boundaries myself.

One theory which this brings up is that the metadata that bgzip adds might be wrong or is not interpreted correctly. In the former case decompressing with `bgzip` should also throw an error. Could you try that? bgzip is part of the tabix package `sudo apt install tabix`.

In order to look at the metadata of that block at that offset, could you run: `pragzip --analyze file.gz | grep -E -C 500 "Compressed Offset   : 234094[0-9]{5} B"`

**Comment by asangphukieo** [comment]
Many thanks for your promt response.
I will try with bgzip command.

This is log file from command "pragzip --analyze file.gz | grep -E -C 500 "Compressed Offset   : 234094[0-9]{5} B"

[pragzip.log](https://github.com/mxmlnkn/pragzip/files/10959142/pragzip.log)

**Comment by mxmlnkn** [comment]
Hopefully fixed with [e9af6eb3](https://github.com/mxmlnkn/indexed_bzip2/commit/e9af6eb30cf69750f181b4861bd94a6bc743d5a5) on master. I cannot reproduce or verify the issue. Please reopen if the problem persists.

**Comment by mxmlnkn** [comment]
I was accidentally able to create a reproducing file. It's basically just the Silesia tarball concatenated 32 times and compressed with bgzip 1.17 using 16 threads. So, something akin to this:

```bash
function m()
{
    if [[ -f Makefile ]]; then
        make -j $( nproc ) "$@"
    elif [[ -f CMakeCache.txt ]]; then
        cmake --build . --parallel $( nproc ) -- "$@"
    fi
}


installPath="$HOME/tmp-tools"

mkdir -p -- "$installPath"


# Install bgzip
(
    version=1.17
    wget -O- "https://github.com/samtools/htslib/releases/download/$version/htslib-$version.tar.bz2" | tar -xj
    mv "htslib-$version" "htslib"
    cd "htslib"
    m -B bgzip
)
export PATH=$installPath/htslib:$PATH


cd ~/indexed_bzip2/build


function getSilesiaTar()
{
    if [[ ! -f 'silesia.zip' || "$( md5sum silesia.zip | sed 's| .*||' )" != 'c240c17d6805fb8a0bde763f1b94cd99' ]]; then
        wget -O 'silesia.zip' 'https://sun.aei.polsl.pl/~sdeor/corpus/silesia.zip'
    fi

    rm -rf silesia/
    mkdir -p silesia && ( cd silesia && unzip ../silesia.zip )
    tar -cf silesia.tar silesia/  # 211957760 B -> 212 MB, 203 MiB, gzip 66 MiB -> compression factor: 3.08
}

getSilesiaTar


parallelization=16

pathToCompress="/tmp/silesia-$(( 2 * parallelization ))x.tar"
for (( i = 0; i < 2 * parallelization; ++i )); do cat silesia.tar; done > "$pathToCompress"
bgzip -c "$pathToCompress" > "$pathToCompress.$compressor" &
```

With version 0.5.0, I can reproduce a similar error:

```bash
git checkout pragzip-v0.5.0
cmake --build . -- pragzip
src/tools/pragzip -P 16 -d -c FAULTY-silesia-32x.tar.bgzip | wc -c
Caught exception:
    Failed to decode deflate block at 2269118464 B 0 b because of: The backreferenced distance lies outside the window buffer!, typeid: St12domain_error
    6778609920
```

Unfortunately, it seems that the bug disappears before the bugfix commit referenced in my last comment:
```bash
git checkout e9af6eb3~1
cmake --build . -- pragzip
src/tools/pragzip -P 16 -d -c FAULTY-silesia-32x.tar.bgzip | wc -c
    Decompressed in total 6782648320 B in 6.51921 s -> 1040.41 MB/s
    6782648320
```

A commit bisection process tells me that this issue has been fixed with d4aa8d4b2fd1d823bc5f2805d83a4b8543fa99be because it works but the commit before does not. Based on the commit message (`[performance] BlockFetcher: Avoid prefetching the same failed offsets multiple times`) it seems indeed somewhat related although I am still surprised that it also is a fix and not just a performance improvement. I might have to take an even closer look to understand the actual cause.
